docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1362 @@
+# KEP-2170: Kubeflow Training V2 API
+
+## Authors
+
+- Andrey Velichkevich - [@andreyvelich](https://github.com/andreyvelich)
+- Yuki Iwai - [@tenzen-y](https://github.com/tenzen-y)
+
+Creation date: 2024-07-16
+
+Google doc: https://bit.ly/3WzjTlw
+
+## Overview
+
+This document discusses the new Kubeflow Training V2 API.
+
+When we built the
+[Kubeflow Training Operator a couple of years ago](https://docs.google.com/document/d/1x1JPDQfDMIbnoQRftDH1IzGU0qvHGSU4W6Jl4rJLPhI/edit?usp=sharing),
+Kubernetes lacked better features to support distributed machine learning (ML) training, such as
+SuccessPolicy and RestartPolicy (FailurePolicy). Recently, the Kubernetes community launched the
+working group Batch, and then the working group actively worked on evolving the batch/v1 `Job` API
+and built [a new `JobSet`](https://github.com/kubernetes-sigs/jobset) API to manage groups of `Jobs`.
+
+This document consolidates efforts for the Cloud Native ML Training between Kubeflow and Kubernetes
+communities.
+
+## Motivation
+
+We often implement features similar to batch/v1 `Job`, such as “suspend”, on the Training Operator
+side since the Training Operator creates blocks of plain Pod and Service for each rank once
+Kubeflow Jobs are created. However, if we continue taking the same approach (re-inventing the wheel),
+the maintenance costs will continue to increase.
+
+It would be better to replace infrastructure layers with `JobSet` to avoid re-inventing the wheel
+and improve the Training Operator
+
+Additionally, introducing `JobSet` as an infrastructure layer would allow us to introduce batch
+workload features such as
+[the PodFailurePolicy](https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-failure-policy)
+and [the PodDisruptionCondition](https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures)
+easily.
+
+Please also see the [Kubernetes JobSet and Kubeflow Training Operator collaboration document](https://docs.google.com/document/d/1C2ev7yRbnMTlQWbQCfX7BCCHcLIAGW2MP9f7YeKl2Ck/edit?usp=sharing).
+
+### User Value
+
+In addition to the above motivation, we will address the following user feedback while implementation:
+
+- Confusion around Workers: https://github.com/kubeflow/training-operator/issues/1790
+- Support batch/v1 `Job` features: https://github.com/kubeflow/training-operator/issues/1718
+- ExitCodes for PodFailurePolicy: https://github.com/kubeflow/training-operator/issues/1749
+- Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906
```

Comment:
There will be three operators for MPI actually.

V1 (which is being deprecated)
V2 (one that uses indexed jobs)
TrainJob.

KubeFlow is hoping to drop V1 soon but not sure when that will happen.

Reply:
We will certainly deprecated MPI V1 Operator as part of this effort.

The question about keeping MPI Operator separate, we need to discuss if we really need to keep it. It is hard to maintain two similar operators.

- We would like to hear from HPC and MPI Operator users if the blueprint idea might be useful for them as well?
- Additionally, If Training Operator can be used for all MPI jobs (ML + HPC) what is the reason to keep MPI Operator separate ?

cc @alculquicondor @tenzen-y

Reply:
MPI is more general purpose instead of focusing on training so IMO it makes more sense to keep it separate.

Reply:
I am not sure if we have enough bandwidth to support MPI Operator and Training Operator separately given that all functionality of MPI Operator will be moved to Training Operator.
I would love to hear some objections from HPC users for this proposal, especially if the idea of re-usable MPI runtimes is not applicable for their use-cases.
cc @kubeflow/wg-training-leads

Reply:
It's hard to make a call (and objection) before the new interfaces exist. An objection would come if functionality is lost by way of moving everything here without considering non-ML use cases of MPI.

Reply:
That makes sense, so maybe after initial implementation we can have this discussion again.

Reply:
IMO, MPI V1 is out definitely. But MPI style launcher mode should be support since DeepSpeed is popular in LLM.

Reply:
In an ideal world, I'd love to have the HPC community aiding/helping with the development of a general MPI operator for Kubernetes that lives outside of Kubeflow. There was some discussion about "moving" MPIOperator to Kubernetes at some point. It sorta died because Kubeflow is part of CNCF but I still see value in building a community supported operator for MPI. 

I think having a limited scope approach for MPI for Kubeflow makes sense. But having MPI under the Kubeflow community has always seemed to confuse HPC/Researchers as I noticed that they don't consider using the MPI Operator from Kubeflow if they are trying to run MPI jobs for fluid dynamics (for example) as Kubeflow is designed for ML usecases.

Reply:
Why HPC community can't contribute to the MPI on Kubernetes with Kubeflow community ? Kubeflow is under CNCF and I don't see any issues for HPC community to work with us to build various HPC runtimes.

>I noticed that they don't consider using the MPI Operator from Kubeflow if they are trying to run MPI jobs for fluid dynamics (for example) as Kubeflow is designed for ML usecases.

I feel that HPC and ML worlds are strongly connected. That is why many folks are looking for HPC technologies to train their models today (e.g. Slurm, MPI, Flux). @vsoch let me know if I am wrong here.

It would be beneficial for both communities if we make Kubernetes more accessible for HPC or AI/ML use-cases.

Reply:
You are entirely correct, and I do advocate that we work together! My organization explicitly has permission to contriute to CNCF, so I can continue to offer my time and attention.

Reply:
> But having MPI under the Kubeflow community has always seemed to confuse HPC/Researchers as I noticed that they don't consider using the MPI Operator from Kubeflow if they are trying to run MPI jobs for fluid dynamics (for example) as Kubeflow is designed for ML usecases.

What do they use instead? My (personal) observation is that it's hard to setup and get working, but I'm also biased to want to use the Flux Operator when we bring MPI into Kubernetes. If you like we can have a discussion about what could be worked on.

Reply:
> Why HPC community can't contribute to the MPI on Kubernetes with Kubeflow community ? Kubeflow is under CNCF and I don't see any issues for HPC community to work with us to build various HPC runtimes.

They can and should! I think having Kubeflow in CNCF will go a long way in getting more people to contribute. 

I think also some teaching on the fact that people can install MPIOperator without installing Kubeflow platform is a necessary point unfortunately. I have had to explain that a lot more than I'd like to admit.

To make it more general, having examples of scientific computing jobs in this repo would be helpful to explain that this project is not just meant for ML.

I am just unsure which repo/project should the development of a general MPI operator fall under. TrainJob is focused on having a pythonic/simple experience for running distributed jobs. Given how vast/complicated MPI can be I am not sure that we can really have that for a general MPI job at the moment. I think having the kubeflow use case covered here would be sufficient.


> What do they use instead? My (personal) observation is that it's hard to setup and get working, but I'm also biased to want to use the Flux Operator when we bring MPI into Kubernetes. If you like we can have a discussion about what could be worked on.

Either they think its impossible to run MPI and Kubernetes and turn to something like Ray or they keep using something like Slurm without Kubernetes seems to be what I've heard. I haven't really worked in too many places that use MPI to be honest.

Reply:
>I think also some teaching on the fact that people can install MPIOperator without installing Kubeflow platform is a necessary point unfortunately. I have had to explain that a lot more than I'd like to admit.

We know this, as we worked a lot in the last 6 months to explain: What is Kubeflow ? 

- https://www.kubeflow.org/docs/started/introduction/#what-are-standalone-kubeflow-components
- https://www.kubeflow.org/docs/started/installing-kubeflow/#standalone-kubeflow-components

Also, we will remove all content from `kubeflow/kubeflow` repo to avoid confusions: https://github.com/kubeflow/kubeflow/issues/7549

> TrainJob is focused on having a pythonic/simple experience for running distributed jobs. 

The main idea of this proposal is to distinguish blueprint from the actual job that user submits. Since many configuration doesn't need to be changed every job submission (schedulers, run policies, node selectors).

As long as we can create HPC runtime, TrainJob can be utilized to start HPC job (although, I understand the naming confusion).

Reply:
> Either they think its impossible to run MPI and Kubernetes and turn to something like Ray or they keep using something like Slurm without Kubernetes seems to be what I've heard. I haven't really worked in too many places that use MPI to be honest.

This sounds like an issue with awareness / advertising. More specifically, the kinds of proxy apps we run in HPC aren't known to the cloud community, so the wires don't cross. I also suspect this - using the "HPC in a box" solutions like HPC Toolkit (Google) or AWS Parallel Cluster instead of Kubernetes solutions.

Reply:
> As long as we can create HPC runtime, TrainJob can be utilized to start HPC job (although, I understand the naming confusion).

I agree - the design is analogous. We have a launcher and workers, and the underlying details of how they hook up is specific to the container and commands provided.

Reply:
> As long as we can create HPC runtime, TrainJob can be utilized to start HPC job (although, I understand the naming confusion).

We could call this DistributedJob?

If MPI is included maybe we can have this be a project for running distributed jobs (training is really just a configuration..).

But I won't make that blocking.

Reply:
Either way, I think I am happy with this discussion as is.

In summary, it seems that there is agreement with forcing out v1 MPIOperator. The Training V2 would include MPI Support.

But there is still an open discussion on what do with v2 mpi operator?

Reply:
>We could call this DistributedJob?

`TrainJob` might give us more flexibility, since maybe users just want to use `TrainJob` to Train Model on single Kubernetes worker and export it somewhere.

>But there is still an open discussion on what do with v2 mpi operator?

Yes, let's followup with the community on what we should do with MPI Operator once we have initial implementation.

Reply:
Sounds good to me then!

Reply:
Sorry for the lately comment.
I'm thinking that we can provide the MPIJob v3 API as the mpi-operator which has the completely same API as the training-operator TrainJob API. As we discussed previously, I'm wondering if we can reuse the API definition by defining the `mlspec.mpi` API in the mpi-operator side, and then import the `mlspec.mpi` API definition to the training-operator repo.

Anyway, I think that it's nice to have a dedicated KEP for the reusable MPI API definition.

Reply:
Yes, let's followup on this after initial implementation. We need to discuss the use-cases from HPC side when they want to have dedicated controller for them since we support all MPI functionality in Training Operator.